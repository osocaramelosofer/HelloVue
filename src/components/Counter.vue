<template>

  <h2>{{ customTitle }}</h2>
  <p>{{counter}} <sup>2</sup> = {{ calculateSquareNumber }}</p>
  <p data-testid="counter">{{ counter }}</p>
  <div>
    <button class="push-button" @click="increaseCounter">+1</button>
    <button class="push-button" v-on:click="decreaseCounter">-1</button>
  </div>


</template>

<script>
export default {
    name: 'Counter', // the name is optional and it's usefull to identify the component in the vuedevtools
    props:{
        title: String,
        start: {
            type: Number,
            default: 100,
            validate( value ){
                return value => 0 
            }
            // required: false
        }
    },

    data(){
        return{
            counter: this.start
        }
    },
    methods:{
        getSquareValue(){
            return this.counter * this.counter
        },
        increaseCounter(){
            this.counter++
        },
        decreaseCounter(){
            this.counter--
        }
    },

    computed:{ // the computed methods always have to return something.
        calculateSquareNumber(){
            return this.counter * this.counter
        },
        customTitle(){
            // return this.title ? this.title : 'Counter'
            return this.title || 'Counter'
        }
    }

}
</script>

<style>
button {
  margin: 40px;
  padding: 0 50px;

  height: 60px;
  border-radius: 15px;
  border: none;

  color: #fff;

  transition: 0.3s ease;

  font-family: "Poppins", sans-serif;
  font-size: 24px;
}
.push-button {
  background-color: #57b894;

  box-shadow: 0 10px #4fa283;
}
.push-button:hover {
  box-shadow: 0 8px #4fa283;
  transform: translateY(1px);
}
.push-button:active {
  background-color: #4a9377;
  box-shadow: 0 5px #407762;
  transform: translateY(4px);
}

</style>