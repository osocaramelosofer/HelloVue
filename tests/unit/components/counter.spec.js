import { shallowMount } from '@vue/test-utils'
import Counter from '@/components/Counter'

describe('Counter Component',  () => {
    let wrapper
    
    beforeEach(() => {
        wrapper = shallowMount( Counter )

    })

    // test('debe de hacer match con el snapshot', () =>{
      
    //     const wrapper = shallowMount( Counter )
    //     expect( wrapper.html() ).toMatchSnapshot()

    // })
    // test('h2 debe de tener el vlaor por defecto "Counter"', () => {
        
    //     const wrapper = shallowMount( Counter )
    //     expect( wrapper.find('h2').exists() ).toBeTruthy() // we check if the component exists
    //     const h2Value = wrapper.find('h2').text()
    //     expect( h2Value ).toBe('Counter')
    // })
    test('el valor por defecto de <p> debe de ser 100', () => {
        
        // ptag
        const pTagValue = wrapper.find('[data-testid="counter"]').text()

        //expect p === 100
        expect( parseInt(pTagValue) ).toEqual(100)
    })

    test('incrementar y decrementar en 1 el valor del contador', async () => {

        const [increaseBtn, decreaseBtn] = wrapper.findAll('button')

        await increaseBtn.trigger('click')
        await increaseBtn.trigger('click')
        await increaseBtn.trigger('click')

        await decreaseBtn.trigger('click')

        const value = wrapper.find('[data-testid="counter"]').text()
        expect(value).toBe('102')

    })
})