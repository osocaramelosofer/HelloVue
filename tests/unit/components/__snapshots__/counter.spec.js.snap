// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Counter Component debe de hacer match con el snapshot 1`] = `
<h2>Counter</h2>
<p>5 <sup>2</sup> = 25</p>
<div><button class="push-button">+1</button><button class="push-button">-1</button></div>
`;
